include/vcpu/arch/intel_x64/vcpu.h

```diff
@@ -420,45 +460,115 @@ class vcpu : public bfvmm::intel_x64::vcpu
         }
     }
 
-    void check_rdmsr()
+    void init_rdmsr()
     {
         check_msr_bitmap();
 
         if (!m_rdmsr) {
             m_rdmsr = std::make_unique<eapis::intel_x64::rdmsr>(
                 m_msr_bitmap.get(), this->exit_handler()
             );
+            m_handlers[vmcs_n::exit_reason::basic_exit_reason::rdmsr] = m_rdmsr.get();
         }
     }
 
-    void check_wrmsr()
+    void init_wrmsr()
     {
         check_msr_bitmap();
 
         if (!m_wrmsr) {
             m_wrmsr = std::make_unique<eapis::intel_x64::wrmsr>(
                 m_msr_bitmap.get(), this->exit_handler()
             );
+            m_handlers[vmcs_n::exit_reason::basic_exit_reason::wrmsr] = m_wrmsr.get();
+        }
+    }
+
+    void init_vpid()
+    {
+        if (!m_vpid) {
+            m_vpid = std::make_unique<eapis::intel_x64::vpid>();
         }
     }
 
 private:
 
+    void init_handlers();
+    void init_exit_handler(vmcs_n::value_type reason);
+
     bool m_is_rdcr3_enabled{false};
     bool m_is_wrcr3_enabled{false};
     bool m_is_rdcr8_enabled{false};
     bool m_is_wrcr8_enabled{false};
 
+    std::unique_ptr<eapis::intel_x64::vpid> m_vpid;
     std::unique_ptr<uint8_t[]> m_msr_bitmap;
 
     std::unique_ptr<eapis::intel_x64::control_register> m_control_register;
     std::unique_ptr<eapis::intel_x64::cpuid> m_cpuid;
     std::unique_ptr<eapis::intel_x64::monitor_trap> m_monitor_trap;
     std::unique_ptr<eapis::intel_x64::mov_dr> m_mov_dr;
     std::unique_ptr<eapis::intel_x64::rdmsr> m_rdmsr;
-    std::unique_ptr<eapis::intel_x64::vpid> m_vpid;
     std::unique_ptr<eapis::intel_x64::wrmsr> m_wrmsr;
+
+    handler_list_t m_handlers;
+    const std::list<vmcs_n::value_type> *m_used_handlers;
 };
 
+const std::list<vmcs_n::value_type>
+vcpu::s_available_handlers{
```

Comment:
Why is s_available_handlers needed? Can we implement this without it? I'd rather not have another malloc just for bookkeeping.